### PR TITLE
Feature/ufs 268 dashboard screen

### PR DESCRIPTION
**Pull request by torchiad**
Using mocked data for now

### CODE REVIEW

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 1]
```diff
@@ -1,11 +1,144 @@
...
+
+    return (
+        <>
+            <Title>Dashboard</Title>
+            <ApplicationContainer>
+                <H4 mb={1}> Applications </H4>
+                <span> In progress ( {props.applications.length} )</span>
+                {props.applications &&
+                    props.applications.length &&
+                    props.applications.map((application: any) => {
```
Comment (torchiad) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Changing....

**Comment by torchiad** [comment]
This is using a mocked response for now

File: src/components/ApplicationDashboard/ApplicationDashboard.test.tsx [hunk 2]
```diff
@@ -4,8 +4,9 @@ import { ApplicationDashboard } from "./index";
 
 describe("ApplicationDashboard", () => {
     it("renders correctly", () => {
-        const tree = renderer.create(<ApplicationDashboard />).toJSON();
+        const tree = renderer
+            .create(<ApplicationDashboard applications={() => ""} />)
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.test.tsx]: What's going on here?

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 3]
```diff
@@ -1,11 +1,158 @@
...
 
 export const ApplicationDashboard: FC<Props> = props => {
-    console.log(props);
-    return <div>Dashboard</div>;
+    const applications =
+        props.applications &&
+        props.applications.listFundingApplications &&
+        props.applications.listFundingApplications.items;
+
+    function daysLeft(date: string | null) {
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Extract this outside the component into a helper that generates the time and suffix (returning the strings, not the JSX. You could do this as an array which you imemdiately destructure). A general rule is to not define functions inside the component unless they need access to hook data. This

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 4]
```diff
@@ -1,11 +1,158 @@
...
+        if (!date) {
+            return <div> No close date</div>;
+        }
+
+        const closeDate = new Date(date);
+        const closeDateTime = closeDate.getTime();
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Check that date has been successfully created

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 5]
```diff
@@ -1,11 +1,158 @@
...
+
+        return (
+            <>
+                <H4 mb={0}>{timeToShow}</H4>
+                <span>{prefixToShow}</span>
+            </>
+        );
+    }
+
+    function friendlyDate(date: string | null) {
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Extract this too

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 6]
```diff
@@ -1,11 +1,158 @@
...
+        );
+    }
+
+    return (
+        <>
+            <Title>Dashboard</Title>
+            <ApplicationContainer>
+                <H4 mb={1}> Applications </H4>
+
+                <span>In progress ({applications && applications.length})</span>
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Change to `{applications ? applications.length : 0 }` or you'll end up with nulls.

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 7]
```diff
@@ -1,11 +1,158 @@
...
+
+                {applications &&
+                    applications.map(application => {
+                        if (!application) {
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx:71]: Don't return loading here. Just return null.

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 8]
```diff
@@ -1,11 +1,158 @@
...
+                                        </div>
+                                        <div>
+                                            Opportunity:
+                                            {application.id}
+                                        </div>
+                                    </GridCol>
+                                    <GridCol setWidth="25%">
+                                        <ApplicationContainerTimeline>
+                                            {daysLeft(application.closeDate)}
+                                            <div>
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Why are you using a div here?

**Review by ascorbic — CHANGES_REQUESTED** [review]

File: src/lib/dateandtime.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        timeToShow = Math.floor(timeLeft / 24);
+        prefixToShow = "Days left";
+    } else if (timeLeft === 1) {
+        prefixToShow = "Hour left";
+    } else if (timeLeft < 0) {
+        timeToShow = 0;
+        prefixToShow = "Ended";
+    }
+
+    const result = { time: timeToShow, prefix: prefixToShow };
```
Comment (ascorbic) [inline src/lib/dateandtime.ts]: No need to assign to a variable. Just return it.

File: src/lib/dateandtime.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
+export function daysLeft(date: string | null) {
```
Comment (ascorbic) [inline src/lib/dateandtime.ts]: Make these arrow functions

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 11]
```diff
@@ -1,11 +1,134 @@
...
+    font-family: ${NTA_LIGHT};
+    text-align: center;
+`;
+
+interface Typed {
+    time?: number;
+    prefix?: string;
+}
+
+function showDaysLeft(closeDate: string | null) {
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: This doesn't need to be a function now

File: src/lib/dateandtime.ts (hunk 9 above)
Comment (ascorbic) [reply src/lib/dateandtime.ts]: return [timetoShow, prefixToShow];

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 12]
```diff
@@ -1,11 +1,134 @@
...
+    text-align: center;
+    border-left: 1px dotted ${GREY_3};
+`;
+
+export const ApplicationContainerTimeline = styled.div`
+    font-family: ${NTA_LIGHT};
+    text-align: center;
+`;
+
+interface Typed {
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: Don't do this here. Type the helper function's return.

**Review by ascorbic — CHANGES_REQUESTED** [review]

File: src/types.d.ts [hunk 13]
```diff
@@ -2,3 +2,10 @@ export interface Organisation {
     ID: string;
     Name: string;
 }
+
+//Helper types
+
```
Comment (ascorbic) [inline src/types.d.ts]: This isn't needed

File: src/components/ApplicationDashboard/ApplicationDashboard.tsx [hunk 14]
```diff
@@ -85,6 +72,15 @@ export const ApplicationDashboard: FC<Props> = props => {
                         if (!application) {
                             return null;
                         }
+
+                        const result: DateandtimeReturnType = {
```
Comment (ascorbic) [inline src/components/ApplicationDashboard/ApplicationDashboard.tsx]: You don't need to spread or type this. You can either just use the result:

`const result = daysLeft(application.closeDate)`

 which is identical to the result you have, or deconstruct it:

`const { timeToShow, prefixToShow } = daysLeft(application.closeDate);`

**Review by ascorbic — CHANGES_REQUESTED** [review]

File: src/lib/dateandtime.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+    const closeDate = new Date(date);
+    const closeDateTime = closeDate.getTime();
+    const today = new Date();
+    const timeToday = today.getTime();
+    const timeleftBig = closeDateTime - timeToday;
+    const timeLeft = Math.floor(timeleftBig / 1000 / 60 / 60);
+
+    let timeToShow = timeLeft;
+    let prefixToShow = "Hours";
```
Comment (ascorbic) [inline src/lib/dateandtime.ts]: Not going to reject the PR for this, but it's a suffix, not a prefix!

**Review by ascorbic — APPROVED** [review]

File: src/lib/dateandtime.ts (hunk 15 above)
Comment (ascorbic) [reply src/lib/dateandtime.ts]: Probably doesn't need the "toShow" bit either.

**Pull request merged**